### ISSUE TITLE
Implementation of Uniformity Loss in Symmetrized InfoNCE?

### ISSUE THREAD

**Issue by dyth** [vivekmyers/contrastive_planning#1]
Hi there, thanks for sharing the code from this v interesting paper!

In the uniformity term of the symmetrized InfoNCE loss

https://github.com/vivekmyers/contrastive_planning/blob/355b02562599d2c04c75f97c148d9c59860696e6/agents/contrastive.py#L56-L57

could I ask why you've multiplied the logits by `(1 - I)`?

Shouldn't the InfoNCE loss contain the positive term in both the numerator and denominator, yielding a uniformity term of 
```python
jax.nn.logsumexp(-pdist, axis=0) + jax.nn.logsumexp(-pdist, axis=1) 
```
?

**Comment by ben-eysenbach** [comment]
The original thinking was that we wanted to exclude the positive examples
(which lie on the diagonal) from the negative loss.
IIRC, some initial experiments showed this worked slightly better, but I
believe that I've ablated this and found that it doesn't actually matter.
Yes, you're right that the version without the (1 - I) term is more
standard. I'd probably recommend using that standard version as the
starting point.

**Comment by dyth** [comment]
Thanks for getting back to me.  Intriguingly there is follow-up work to InfoNCE called DCL (https://arxiv.org/abs/2110.06848) that removes the positive terms from the denominator in a similar way.  They show that this improves the performance of contrastive learning, especially at small batch sizes.
